browser/tor/tor_profile_service_impl.cc

```diff
@@ -46,42 +46,44 @@ void TorProfileServiceImpl::ReLaunchTor(const TorConfig& config) {
   tor_launcher_factory_->ReLaunchTorProcess(config);
 }
 
-void TorProfileServiceImpl::SetNewTorCircuitOnIOThread(
-    const scoped_refptr<net::URLRequestContextGetter>& getter,
-    std::string host) {
-  DCHECK_CURRENTLY_ON(BrowserThread::IO);
-  const TorConfig tor_config = tor_launcher_factory_->GetTorConfig();
-  if (tor_config.empty())
-    return;
-  auto* proxy_resolution_service =
-    getter->GetURLRequestContext()->proxy_resolution_service();
-  DCHECK(proxy_resolution_service);
-  TorProxyConfigService::TorSetProxy(proxy_resolution_service,
-                                     tor_config.proxy_string(),
-                                     host,
-                                     &tor_proxy_map_,
-                                     true);
+void TorProfileServiceImpl::OnSetNewTorCircuitComplete(bool success) {
+  if (tor_circuit_callback_)
+      std::move(tor_circuit_callback_).Run(success);
+}
+
+void TorProfileServiceImpl::OnProxyLookupComplete(
+    int32_t net_error,
+    const base::Optional<net::ProxyInfo>& proxy_info) {
+  bool success = proxy_info.has_value() && !proxy_info->is_direct();
+  if (tor_circuit_callback_) {
+    std::move(tor_circuit_callback_).Run(success);
+    binding_.Close();
+  }
 }
 
 void TorProfileServiceImpl::SetNewTorCircuit(const GURL& request_url,
-                                             const base::Closure& callback) {
-  std::string isolation_key = CircuitIsolationKey(request_url);
-  if (isolation_key.empty())
-    return;
+                                             NewTorCircuitCallback callback) {
   auto* storage_partition =
     BrowserContext::GetStoragePartitionForSite(profile_, request_url, false);
 
-  net::URLRequestContextGetter* url_request_context_getter =
-    storage_partition->GetURLRequestContext();
-  DCHECK(url_request_context_getter);
-
-  base::PostTaskWithTraitsAndReply(
-      FROM_HERE, {BrowserThread::IO},
-      base::Bind(&TorProfileServiceImpl::SetNewTorCircuitOnIOThread,
-                 base::Unretained(this),
-                 base::WrapRefCounted(url_request_context_getter),
-                 isolation_key),
-    callback);
+  GURL::Replacements replacements;
+  replacements.SetRef("NewTorCircuit",
+                      url::Component(0, strlen("NewTorCircuit")));
+  GURL url = request_url.ReplaceComponents(replacements);
+  tor_circuit_callback_ = std::move(callback);
+
+  DCHECK_CURRENTLY_ON(content::BrowserThread::UI);
```

Comment:
typically such checks are placed in the very beginning of a method

Reply:
fixed in 51b3f3c